src/subdomains/core/custody/services/custody.service.ts

```diff
@@ -0,0 +1,169 @@
+    if (ref) dto.usedRef ??= ref.ref;
+
+    const wallet =
+      (await this.walletService.getByIdOrName(undefined, dto.wallet)) ?? (await this.walletService.getDefault());
+
+    const addressIndex = await this.userService.getNexCustodyIndex();
+    const custodyWallet = EvmUtil.createWallet(Config.blockchain.evm.custodyAccount(addressIndex));
+    const signature = await custodyWallet.signMessage(Config.auth.signMessageGeneral + custodyWallet.address);
+
+    const account = await this.userDataService.getUserData(accountId);
+    if (!account) throw new NotFoundException('User not found');
+
+    const custodyUser = await this.userService.createUser(
+      {
+        address: custodyWallet.address,
+        signature,
+        usedRef: dto.usedRef,
+        ip: userIp,
+        origin: ref?.origin,
+        wallet,
+        userData: account,
+        custodyAddressType: dto.addressType,
+        custodyAddressIndex: addressIndex,
+        role: UserRole.CUSTODY,
+      },
+      dto.specialCode,
+    );
+
+    return { accessToken: this.authService.generateUserToken(custodyUser, userIp) };
+  }
+
+  async createStep(
+    order: CustodyOrder,
+    index: number,
+    command: string,
+    context: CustodyOrderStepContext,
+  ): Promise<CustodyOrderStep> {
+    const orderStep = this.custodyOrderStepRepo.create({
+      order,
+      index,
+      command,
+      context,
+    });
+    return this.custodyOrderStepRepo.save(orderStep);
+  }
+
+  async startNextStep(step: CustodyOrderStep): Promise<void> {
+    const nextIndex = step.index + 1;
+    const order = step.order;
+    const nextStep = OrderConfig[order.type][nextIndex];
+
+    if (nextStep) {
+      await this.createStep(order, nextIndex, nextStep.command, nextStep.context);
+    } else {
+      const asset = order.inputAsset ?? order.outputAsset;
+      const amount = order.inputAmount ?? -order.outputAmount;
+
+      await this.custodyOrderRepo.update(...order.complete());
+
+      const custodyBalance = await this.custodyBalanceRepo.findOneBy({
+        asset: { id: asset.id },
+        user: { id: order.user.id },
+      });
+
+      if (!custodyBalance) {
+        await this.createCustodyBalance(amount, order.user, asset);
+      } else {
+        const { deposit, withdrawal } = await this.custodyOrderRepo
+          .createQueryBuilder('custodyOrder')
+          .select('SUM(custodyOrder.inputAmount)', 'deposit')
+          .addSelect('SUM(custodyOrder.outputAmount)', 'withdrawal')
+          .where('custodyOrder.userId = :id', { id: order.user.id })
+          .andWhere('custodyOrder.status = :status', { status: CustodyOrderStatus.COMPLETED })
+          .andWhere('(custodyOrder.inputAssetId = :asset OR custodyOrder.outputAssetId = :asset)', {
+            asset: asset.id,
+          })
+          .getRawOne<{ deposit: number; withdrawal: number }>();
+
+        await this.updateCustodyBalance(custodyBalance, deposit - withdrawal + amount);
+      }
```

Comment:
Hm, this will update the balance for deposit transactions as soon as the user confirmed it 🤔  for withdraws we probably need to update the balance on order start, for deposits on order end.